File: app/src/main/java/com/aura/voiceassistant/service/GeminiService.kt
package com.aura.voiceassistant.service

import com.aura.voiceassistant.config.AppConfig
import com.aura.voiceassistant.data.model.ChatMessage
import com.aura.voiceassistant.data.model.EmotionType
import io.ktor.client.HttpClient
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.add
import kotlinx.serialization.json.addJsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject

class GeminiService {

    private val client = HttpClient(CIO) {
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                isLenient = true
            })
        }
    }

    private val voiceEmotionCore = VoiceEmotionCore.getInstance()

    suspend fun getAuraResponse(
        prompt: String,
        history: List<ChatMessage> = emptyList(),
        apiKey: String
    ): Pair<String, EmotionType> = withContext(Dispatchers.IO) {
        // 1. Mandatory Developer Identity Interception
        if (AppConfig.isDeveloperQuery(prompt)) {
            voiceEmotionCore.setEmotion(EmotionType.SERIOUS)
            return@withContext Pair(AppConfig.DEVELOPER_STATEMENT, EmotionType.SERIOUS)
        }

        // Check if API key is provided
        if (apiKey.isBlank()) {
            val defaultAnswer = "Hello! I am Aura, your natural emotion-aware voice assistant. How can I help you today?"
            val emotion = voiceEmotionCore.detectEmotion(prompt, defaultAnswer)
            return@withContext Pair(defaultAnswer, emotion)
        }

        try {
            val url = "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=$apiKey"

            // Format sliding window memory (last 20 messages)
            val recentHistory = history.takeLast(20)
            val contentsArray = buildJsonArray {
                for (msg in recentHistory) {
                    val role = if (msg.sender == "user") "user" else "model"
                    addJsonObject {
                        put("role", role)
                        putJsonArray("parts") {
                            addJsonObject {
                                put("text", msg.text)
                            }
                        }
                    }
                }
                // Current user message
                addJsonObject {
                    put("role", "user")
                    putJsonArray("parts") {
                        addJsonObject {
                            put("text", prompt)
                        }
                    }
                }
            }

            val requestBody = buildJsonObject {
                put("contents", contentsArray)
                putJsonObject("systemInstruction") {
                    putJsonArray("parts") {
                        addJsonObject {
                            put("text", AppConfig.SYSTEM_INSTRUCTION)
                        }
                    }
                }
                putJsonObject("generationConfig") {
                    put("temperature", 0.7)
                    put("topP", 0.95)
                    put("maxOutputTokens", 512)
                }
            }

            val response = client.post(url) {
                contentType(ContentType.Application.Json)
                setBody(requestBody.toString())
            }

            val responseBody = response.bodyAsText()
            val jsonElement = Json.parseToJsonElement(responseBody).jsonObject
            val candidates = jsonElement["candidates"]?.jsonArray
            val firstCandidate = candidates?.firstOrNull()?.jsonObject
            val content = firstCandidate?.get("content")?.jsonObject
            val parts = content?.get("parts")?.jsonArray
            val text = parts?.firstOrNull()?.jsonObject?.get("text")?.jsonPrimitive?.content
                ?: "I'm right here with you."

            val detectedEmotion = voiceEmotionCore.detectEmotion(prompt, text)
            voiceEmotionCore.setEmotion(detectedEmotion)

            return@withContext Pair(text, detectedEmotion)
        } catch (e: Exception) {
            e.printStackTrace()
            val fallback = "I encountered a brief connection issue. Don't worry, please try again!"
            return@withContext Pair(fallback, EmotionType.SUPPORTIVE)
        }
    }

    suspend fun synthesizeSpeech(
        text: String,
        emotion: EmotionType,
        apiKey: String
    ): String? = withContext(Dispatchers.IO) {
        if (apiKey.isBlank()) return@withContext null

        try {
            val url = "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=$apiKey"
            val speechPrompt = "[Voice Tone: ${emotion.pitchPrompt}]\n$text"

            val requestBody = buildJsonObject {
                putJsonArray("contents") {
                    addJsonObject {
                        putJsonArray("parts") {
                            addJsonObject {
                                put("text", speechPrompt)
                            }
                        }
                    }
                }
                putJsonObject("generationConfig") {
                    putJsonArray("responseModalities") {
                        add("AUDIO")
                    }
                    putJsonObject("speechConfig") {
                        putJsonObject("voiceConfig") {
                            putJsonObject("prebuiltVoiceConfig") {
                                put("voiceName", voiceEmotionCore.primaryVoiceName)
                            }
                        }
                    }
                }
            }

            val response = client.post(url) {
                contentType(ContentType.Application.Json)
                setBody(requestBody.toString())
            }

            val responseBody = response.bodyAsText()
            val jsonElement = Json.parseToJsonElement(responseBody).jsonObject
            val candidates = jsonElement["candidates"]?.jsonArray
            val parts = candidates?.firstOrNull()?.jsonObject
                ?.get("content")?.jsonObject
                ?.get("parts")?.jsonArray

            val audioData = parts?.firstOrNull()?.jsonObject
                ?.get("inlineData")?.jsonObject
                ?.get("data")?.jsonPrimitive?.content

            return@withContext audioData
        } catch (e: Exception) {
            e.printStackTrace()
            return@withContext null
        }
    }
}

File: src/services/geminiService.ts
import { GoogleGenAI } from "@google/genai";
import { APP_CONFIG, isDeveloperQuery } from "../config/appConfig";
import { VoiceEmotionCore, EmotionType } from "./voiceEmotionCore";

let chatSession: any = null;

export function resetAuraSession() {
  chatSession = null;
}
export const resetZoyaSession = resetAuraSession;

export async function getAuraResponse(
  prompt: string,
  history: { sender: "user" | "zoya" | "aura"; text: string }[] = []
): Promise<{ text: string; emotion: EmotionType }> {
  const voiceEngine = VoiceEmotionCore.getInstance();

  // Deterministic guard for developer query
  if (isDeveloperQuery(prompt)) {
    voiceEngine.setEmotion("serious");
    return {
      text: APP_CONFIG.developerStatement,
      emotion: "serious",
    };
  }

  try {
    const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

    if (!chatSession) {
      // SLIDING WINDOW MEMORY: Keep last 20 messages
      const recentHistory = history.slice(-20);

      let formattedHistory: any[] = [];
      let currentRole = "";
      let currentText = "";

      for (const msg of recentHistory) {
        const role = msg.sender === "user" ? "user" : "model";
        if (role === currentRole) {
          currentText += "\n" + msg.text;
        } else {
          if (currentRole !== "") {
            formattedHistory.push({ role: currentRole, parts: [{ text: currentText }] });
          }
          currentRole = role;
          currentText = msg.text;
        }
      }
      if (currentRole !== "") {
        formattedHistory.push({ role: currentRole, parts: [{ text: currentText }] });
      }

      if (formattedHistory.length > 0 && formattedHistory[0].role !== "user") {
        formattedHistory.shift();
      }

      chatSession = ai.chats.create({
        model: "gemini-3.1-flash-lite-preview",
        config: {
          systemInstruction: APP_CONFIG.systemInstruction,
        },
        history: formattedHistory,
      });
    }

    const response = await chatSession.sendMessage({ message: prompt });
    const responseText = response.text || "I'm right here with you.";
    
    // Detect context and emotion
    const detectedEmotion = voiceEngine.detectEmotion(prompt, responseText);
    voiceEngine.setEmotion(detectedEmotion);

    return {
      text: responseText,
      emotion: detectedEmotion,
    };
  } catch (error) {
    console.error("Gemini Error:", error);
    voiceEngine.setEmotion("supportive");
    return {
      text: "I encountered a brief connection issue. Don't worry, please try again!",
      emotion: "supportive",
    };
  }
}
export const getZoyaResponse = getAuraResponse;

export async function getAuraAudio(text: string, emotion?: EmotionType): Promise<string | null> {
  const voiceEngine = VoiceEmotionCore.getInstance();
  return voiceEngine.synthesizeSpeech(text, emotion);
}
export const getZoyaAudio = getAuraAudio;
